File: tst/layers.c
#include "pico.h"
#include <assert.h>
#include <string.h>
#include "../check.h"

int main (void) {
    pico_init(1);

    // create bg layer (32x32) - detached (up=NULL)
    puts("create background layer");
    pico_layer_empty(NULL, "background", (Pico_Abs_Dim){32, 32}, NULL);

    // create ui layer (48x48) - detached
    puts("create ui layer");
    pico_layer_empty(NULL, "ui", (Pico_Abs_Dim){48, 48}, NULL);

    // draw on bg layer (red background)
    puts("draw on background layer directly");
    pico_set_effect_color("background", (Pico_Color){0x80, 0x00, 0x00, 0xFF});
    pico_output_clear("background");
    pico_set_pencil_color("background", (Pico_Color){0xFF, 0xFF, 0xFF, 0xFF});
    pico_output_draw_rect("background", &(Pico_Rel_Rect){ '%', {0.5, 0.5, 0.5, 0.5}, PICO_ANCHOR_C, NULL });
    // composite onto root
    pico_output_clear("root");
    pico_output_draw_layer("root", "background", &(Pico_Rel_Rect){ '%', {0.5, 0.5, 1, 1}, PICO_ANCHOR_C, NULL });
    _pico_check("layers-01");

    // draw on ui layer (blue background)
    puts("draw on ui layer directly");
    pico_set_effect_color("ui", (Pico_Color){0x00, 0x00, 0x80, 0xFF});
    pico_output_clear("ui");
    pico_set_pencil_color("ui", (Pico_Color){0x00, 0xFF, 0x00, 0xFF});
    pico_output_draw_rect("ui", &(Pico_Rel_Rect){ '%', {0.5, 0.5, 0.5, 0.5}, PICO_ANCHOR_C, NULL });
    // composite onto root
    pico_output_clear("root");
    pico_output_draw_layer("root", "ui", &(Pico_Rel_Rect){ '%', {0.5, 0.5, 1, 1}, PICO_ANCHOR_C, NULL });
    _pico_check("layers-02");

    // composite both layers onto root
    puts("composite layers onto root");
    pico_set_effect_color("root", (Pico_Color){0x00, 0x00, 0x00, 0xFF});
    pico_output_clear("root");
    pico_output_draw_layer("root", "background", &(Pico_Rel_Rect){ '%', {1.0/3, 1.0/3, 1.0/3, 1.0/3}, PICO_ANCHOR_C, NULL });
    pico_output_draw_layer("root", "ui", &(Pico_Rel_Rect){ '%', {2.0/3, 2.0/3, 1.0/3, 1.0/3}, PICO_ANCHOR_C, NULL });
    _pico_check("layers-03");

    // present works
    puts("present works");
    pico_output_present();

    // pico_layer_empty reuse (content preserved)
    puts("layer_empty reuse");
    pico_layer_empty_mode('=', NULL, "reuse", (Pico_Abs_Dim){32, 32}, NULL);
    pico_set_effect_color("reuse", (Pico_Color){0x00, 0x80, 0x00, 0xFF});
    pico_output_clear("reuse");
    pico_set_pencil_color("reuse", (Pico_Color){0xFF, 0xFF, 0x00, 0xFF});
    pico_output_draw_rect("reuse", &(Pico_Rel_Rect){'%', {0.5, 0.5, 0.5, 0.5}, PICO_ANCHOR_C, NULL});
    pico_layer_empty_mode('=', NULL, "reuse", (Pico_Abs_Dim){64, 64}, NULL);
    pico_set_effect_color("root", (Pico_Color){0x00, 0x00, 0x00, 0xFF});
    pico_output_clear("root");
    pico_output_draw_layer("root", "reuse", &(Pico_Rel_Rect){'%', {0.5, 0.5, 1, 1}, PICO_ANCHOR_C, NULL});
    _pico_check("layers-04");

    // pico_layer_pixmap with name (reuse)
    puts("layer_pixmap with name");
    static Pico_Color buf1[4] = {
        {255, 0, 0, 255}, {0, 255, 0, 255},
        {0, 0, 255, 255}, {255, 255, 0, 255}
    };
    pico_layer_pixmap_mode('=', NULL, "mybuf", (Pico_Abs_Dim){2, 2}, buf1);
    pico_layer_pixmap_mode('=', NULL, "mybuf", (Pico_Abs_Dim){2, 2}, buf1);

    // pico_layer_pixmap with different name
    puts("layer_pixmap different name");
    pico_layer_pixmap_mode('=', NULL, "xxx", (Pico_Abs_Dim){2, 2}, buf1);
    pico_layer_pixmap_mode('=', NULL, "xxx", (Pico_Abs_Dim){2, 2}, buf1);

    // draw pixmap layer
    puts("draw pixmap layer");
    pico_set_effect_color("root", (Pico_Color){0x00, 0x00, 0x00, 0xFF});
    pico_output_clear("root");
    pico_output_draw_layer("root", "mybuf", &(Pico_Rel_Rect){'%', {0.5, 0.5, 0.5, 0.5}, PICO_ANCHOR_C, NULL});
    _pico_check("layers-05");

    pico_init(0);
    return 0;
}

File: lua/tst/layers.lua
require 'pico.check'

pico.init(true)

-- get_layer returns nil (main layer)
print("get_layer returns nil initially")
local layer = pico.get.layer()
assert(layer == "world")

-- set_layer(nil) switches to main
print("set_layer(nil) keeps main layer")
pico.set.layer("world")
layer = pico.get.layer()
assert(layer == "world")

-- create bg layer (32x32)
print("create and switch to layer")
pico.layer.empty { mode='!', key="background", dim={'!', w=32, h=32} }
pico.set.layer("background")
layer = pico.get.layer()
assert(layer == "background")

-- create ui layer (48x48)
print("switch to another layer")
pico.layer.empty { mode='!', key="ui", dim={'!', w=48, h=48} }
pico.set.layer("ui")
layer = pico.get.layer()
assert(layer == "ui")

-- draw on bg layer (red background)
print("draw on layer (no auto-present)")
pico.set.layer("background")
pico.set.effect { color={'!', r=0x80, g=0x00, b=0x00} }
pico.output.clear()
pico.output.draw.rect({'%', x=0.5, y=0.5, w=0.5, h=0.5, anchor='C'})
pico.set.layer("world")
pico.output.clear()
pico.output.draw.layer("background", {'%', x=0.5, y=0.5, w=1, h=1, anchor='C'})
pico.check("layers-01")

-- draw on ui layer (blue background)
print("draw on ui layer")
pico.set.layer("ui")
pico.set.effect { color={'!', r=0x00, g=0x00, b=0x80} }
pico.output.clear()
pico.set.pencil { color='green' }
pico.output.draw.rect({'%', x=0.5, y=0.5, w=0.5, h=0.5, anchor='C'})
pico.set.layer("world")
pico.output.clear()
pico.output.draw.layer("ui", {'%', x=0.5, y=0.5, w=1, h=1, anchor='C'})
pico.check("layers-02")

-- switch back to main
print("switch back to main")
pico.set.layer("world")
layer = pico.get.layer()
assert(layer == "world")

-- composite layers onto main
print("draw layers onto main")
pico.set.effect { color='black' }
pico.output.clear()
pico.output.draw.layer("background", {'%', x=1.0/3, y=1.0/3, h=1.0/3, anchor='C'})
pico.output.draw.layer("ui", {'%', x=2.0/3, y=2.0/3, w=1.0/3, anchor='C'})
pico.check("layers-03")

-- present works on main
print("present works on main")
pico.output.present()

-- rect-as-dim shortcut: 4th arg with x/y sets both dim and scene.target
print("layer.empty: rect-as-dim shortcut sets dim + target")
do
    -- world is 100x100; '%' w=0.5,h=0.5 -> 50x50 layer
    local r = {'%', x=0.25, y=0.75, w=0.5, h=0.5, anchor='C'}
    pico.layer.empty { up="world", key="rect_shortcut", clear=true, target=r }
    pico.set.layer("rect_shortcut")
    local s = pico.get.scene()
    -- dim derived from w/h (rect's '%' resolved against parent world)
    assert(s.dim.w == 50 and s.dim.h == 50)
    -- target set to the full rect (mode + x/y/w/h preserved)
    assert(s.target[1]  == '%')
    assert(s.target.x == 0.25 and s.target.y == 0.75)
    assert(s.target.w == 0.5  and s.target.h == 0.5)
    pico.set.layer("world")
end

-- trailing rect on other creators: pixmap with target rect set in one call
print("layer.pixmap: trailing rect sets scene.target")
do
    local r = {'%', x=0.5, y=0.5, w=0.25, h=0.25, anchor='C'}
    local px = {{ {'!',r=0xFF,g=0,b=0}, {'!',r=0,g=0xFF,b=0} },
                { {'!',r=0,g=0,b=0xFF}, {'!',r=0xFF,g=0xFF,b=0} }}
    pico.layer.pixmap { up="world", key="px_with_target", pix=px, target=r }
    pico.set.layer("px_with_target")
    local s = pico.get.scene()
    assert(s.target[1] == '%')
    assert(s.target.x == 0.5 and s.target.y == 0.5)
    assert(s.target.w == 0.25 and s.target.h == 0.25)
    pico.set.layer("world")
end

pico.init(false)
